### PR TITLE
Portable proxy deployment utility

### PR DESCRIPTION
**Pull request by danistrebel**
- Proxy utility for Apigee Edge and hybrid/X

- [x] I have run all the tests locally and they all pass.
- [x] I have followed the relevant style guide for my changes.

**CC:** @apigee-devrel-reviewers

### CODE REVIEW

**Comment by laughingbiscuit** [comment]
silly question - why is this better than a two line little script that does something like...

```
set -e
curl -sSL https://github.com/apigee/devrel/tools/.../pom.xml # get the perfect pom
mvn install $@ # pass args to mvn install
```

... and adding extra params for proxy name and basepath in the 'perfect pom'? My concern is the separation of logic between the maven deploy plugin and this script. it would be easier to maintain if it was a 'super thin' wrapper without if statements and regex matching. Thoughts?

**Comment by danistrebel** [comment]
@laughingbiscuit fair question. The idea was based on the following need:
* The solutions is a wrapper around the maven plugin where the underlying plugin could be swapped out for anything else in the future.
* We want to abstract the complexities around the different configurations for Edge and hybrid/X
* Users should not have to worry about installing maven and java just for the purpose of running a Proxy deployment.
* Base path overrides is not a feature of the maven plugin (only apigeetool)

**Comment by seymen** [comment]
Where this will come short is when people also depend on the config plugin. Is there anything we can do for that, I wonder?

File: tools/portable-proxy-deployer/pom.xml [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+			</properties>
+		</profile>
+		<profile>
+			<!-- Apigee X and hybrid -->
+			<id>googleapi</id>
+			<properties>
+				<apigee.plugin.version>2.1.0</apigee.plugin.version>
+				<apigee.hosturl>https://apigee.googleapis.com</apigee.hosturl>
+				<apigee.bearer>${token}</apigee.bearer> <!-- this takes precedence over service account file -->
+				<apigee.serviceaccount.file>${sa}</apigee.serviceaccount.file>
```
Comment (laughingbiscuit) [inline tools/portable-proxy-deployer/pom.xml:77]: drop this

File: tools/portable-proxy-deployer/deploy.sh [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  esac
+done
+
+if [[ -z "$token" && (-z "$username"  || -z "$password") ]]; then
+    echo "[FATAL] required either -t (GCP access token) or -u and -p (Edge username and password)"
+    exit 1
+fi
+
+# Make temp 'deploy-me' directory to keep things clean
+mkdir -p ./deploy-me
```
[the remark this answers is not in the document]
Comment (laughingbiscuit) [reply tools/portable-proxy-deployer/deploy.sh:81]: make sure we have cleaned up before we start

**Review by laughingbiscuit — APPROVED** [review]

File: tools/portable-proxy-deployer/deploy.sh [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+fi
+
+# Determine Proxy name
+proxy_name_in_bundle="$(xpath -q -e 'string(//APIProxy/@name)' ./apiproxy/*.xml)"
+api_name=${api_name:=$proxy_name_in_bundle}
+
+# (optional) Override base path
+if [ -n "$base_path" ]; then
+    echo "[INFO] Setting base path: $base_path"
+    sed -i.bak "s|<BasePath>.*</BasePath>|<BasePath>$base_path<\/BasePath>|g" ./apiproxy/proxies/*.xml
```
[the remark this answers is not in the document]
Comment (danistrebel) [reply tools/portable-proxy-deployer/deploy.sh]: deferred

File: tools/portable-proxy-deployer/deploy.sh [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    cp -r "$git_repo/$git_path" .
+elif [ -n "$directory" ]; then
+    echo "[INFO] using local directory: $directory"
+    (cd .. && cp -r "$directory" ./deploy-me/apiproxy)
+else
+    cp -r ../apiproxy ./apiproxy
+fi
+
+# Determine Proxy name
+proxy_name_in_bundle="$(xpath -q -e 'string(//APIProxy/@name)' ./apiproxy/*.xml)"
```
[the remark this answers is not in the document]
Comment (danistrebel) [reply tools/portable-proxy-deployer/deploy.sh]: xmllint -> defer

**Pull request merged**